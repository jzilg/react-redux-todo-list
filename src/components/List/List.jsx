import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import Todo from '../Todo'
import './list.scss'

const List = ({
    todos,
    addTodo,
    saveTodo,
    removeTodo,
    isLoading,
}) => {
    /**
     * @returns {number}
     */
    const createNewId = () => {
        if (!todos.size) {
            return 0
        }
        const ids = todos.map(todo => todo.id)
        const highestId = Math.max(...ids)
        return highestId + 1
    }

    const addNewTodo = () => {
        const id = createNewId()
        addTodo(id)
    }

    const todoElements = todos.map(todo => (
        <li key={todo.id} styleName="list-element">
            <Todo
                todo={todo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        </li>
    ))

    return (
        <div className="container">
            <h1>
                Todo List
            </h1>
            <ul styleName="list">
                {todoElements}
            </ul>
            <button
                type="button"
                styleName="add-todo-btn"
                title="Add Todo"
                onClick={addNewTodo}
                disabled={isLoading}
            >
                +
            </button>
        </div>
    )
}

List.propTypes = {
    todos: ImmutablePropTypes.list.isRequired,
    addTodo: PropTypes.func.isRequired,
    saveTodo: PropTypes.func.isRequired,
    removeTodo: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
}

export default List
