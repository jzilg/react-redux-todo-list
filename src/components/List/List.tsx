import React, { ReactElement } from 'react'
import TodoType from '../../interfaces/todo.interface'
import Todo from '../Todo'
import style from './list.scss'

interface ListProps {
    todos: TodoType[],
    addTodo: Function,
    saveTodo: Function,
    removeTodo: Function,
    isLoading: boolean,
}

const List = ({
    todos,
    addTodo,
    saveTodo,
    removeTodo,
    isLoading,
}: ListProps): ReactElement<{}> => {
    const createNewId = (): number => {
        if (!todos.length) {
            return 0
        }
        const ids = todos.map(todo => todo.id)
        const highestId = Math.max(...ids)
        return highestId + 1
    }

    const addNewTodo = (): void => {
        const id = createNewId()
        addTodo(id)
    }

    const todoElements = todos.map(todo => (
        <li key={todo.id} className={style['list-element']}>
            <Todo
                todo={todo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        </li>
    ))

    return (
        <div className="container">
            <h1>
                Todo List
            </h1>
            <ul className={style.list}>
                {todoElements}
            </ul>
            <button
                type="button"
                className={style['add-todo-btn']}
                title="Add Todo"
                onClick={addNewTodo}
                disabled={isLoading}
            >
                +
            </button>
        </div>
    )
}

export default List
