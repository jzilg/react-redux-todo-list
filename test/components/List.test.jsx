import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { shallowToJson } from 'enzyme-to-json'
import expect from 'expect'
import List from '../../src/components/List'

Enzyme.configure({ adapter: new Adapter() })

describe('List', () => {
    const saveTodo = () => {}
    const removeTodo = () => {}
    const isLoading = false
    const todos = [
        {
            id: 1,
            name: 'Papa anrufen',
            schedule: 4,
            lastEvent: '2018-06-03',
        },
        {
            id: 2,
            name: 'Superfit',
            schedule: 3,
            lastEvent: '2018-05-11',
        },
        {
            id: 3,
            name: 'Mama anrufen',
            schedule: 7,
            lastEvent: '2018-05-12',
        },
    ]

    it('should render correcty one todo', () => {
        const addTodo = () => {}
        const todo = todos.slice().splice(0, 1)
        const component = shallow((
            <List
                todos={todo}
                addTodo={addTodo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })

    it('should render correcty more todos right sorted', () => {
        const addTodo = () => {}
        const component = shallow((
            <List
                todos={todos}
                addTodo={addTodo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })

    it('should call add todo if button is clicked', () => {
        const addTodo = jest.fn()
        const component = shallow((
            <List
                todos={todos}
                addTodo={addTodo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        ))
        component.find('button').simulate('click')
        expect(addTodo.mock.calls.length).toBe(1)
    })
})
