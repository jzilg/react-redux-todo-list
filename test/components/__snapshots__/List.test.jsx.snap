// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render correctly more todos 1`] = `
<div
  className="container"
>
  <h1>
    Todo List
  </h1>
  <ul
    styleName="list"
  >
    <li
      key="1"
      styleName="list-element"
    >
      <Todo
        isLoading={false}
        removeTodo={[Function]}
        saveTodo={[Function]}
        todo={
          Object {
            "id": 1,
            "lastEvent": "2018-06-03",
            "name": "Papa anrufen",
            "schedule": 4,
          }
        }
      />
    </li>
    <li
      key="2"
      styleName="list-element"
    >
      <Todo
        isLoading={false}
        removeTodo={[Function]}
        saveTodo={[Function]}
        todo={
          Object {
            "id": 2,
            "lastEvent": "2018-05-11",
            "name": "Superfit",
            "schedule": 3,
          }
        }
      />
    </li>
    <li
      key="3"
      styleName="list-element"
    >
      <Todo
        isLoading={false}
        removeTodo={[Function]}
        saveTodo={[Function]}
        todo={
          Object {
            "id": 3,
            "lastEvent": "2018-05-12",
            "name": "Mama anrufen",
            "schedule": 7,
          }
        }
      />
    </li>
  </ul>
  <button
    disabled={false}
    onClick={[Function]}
    styleName="add-todo-btn"
    title="Add Todo"
    type="button"
  >
    +
  </button>
</div>
`;

exports[`List should render correctly one todo 1`] = `
<div
  className="container"
>
  <h1>
    Todo List
  </h1>
  <ul
    styleName="list"
  >
    <li
      key="1"
      styleName="list-element"
    >
      <Todo
        isLoading={false}
        removeTodo={[Function]}
        saveTodo={[Function]}
        todo={
          Object {
            "id": 1,
            "lastEvent": "2018-06-03",
            "name": "Papa anrufen",
            "schedule": 4,
          }
        }
      />
    </li>
  </ul>
  <button
    disabled={false}
    onClick={[Function]}
    styleName="add-todo-btn"
    title="Add Todo"
    type="button"
  >
    +
  </button>
</div>
`;
